src/transaction_builder.js

```diff
@@ -210,8 +212,18 @@ function prepareInput (input, kpPubKey, redeemScript, hashType) {
     // pay-to-scriptHash is not possible without a redeemScript
     if (input.prevOutType === 'scripthash') throw new Error('PrevOutScript is P2SH, missing redeemScript')
 
+    if (!input.pubKeys || input.pubKeys.length === 0) {
+      // maybe we can derive more information about the script
+      // now that we have a kpPubKey?
+      expanded = expandOutput(input.prevOutScript, input.prevOutType, kpPubKey)
+      input.pubKeys = expanded.pubKeys
+      input.signatures = expanded.signatures
```

Comment:
@runn1ng indeed!
I'm hoping to continue improving that,  and have the flow clearly outlined and probably more comments added.

*  `prepareInput` is done only if the input was not previously "prepared" when `addInput` was called.
* `expandInput` is done when a `scriptSig` is already present,  aka from `fromTransaction`.  This script also calls itself recursively if you supplied a `redeemScript`.
* `expandOutput` is done when we have a `prevOutScript` (aka, the previous transactions output) and are attempting to derive information from that script about how we might sign it.

`expandOutput` is basically useless except in the raw multisig or raw pubkey output script cases... which are very rare,  but,  if it is a pubkeyhash output... and you have your public key handy,  you can check if its pubkey hash matches which means you would be able to sign it, and hence,  it derives what is needed for that!

I'm sure it'll get cleaner,  but that is mostly the gist, with a few exceptions for `scripthash` outputs.